pkg/codegen/nodejs/gen_program.go

```diff
@@ -362,6 +362,9 @@ func outputRequiresAsyncMain(ov *pcl.OutputVariable) bool {
 func resourceTypeName(r *pcl.Resource) (string, string, string, hcl.Diagnostics) {
 	// Compute the resource type from the Pulumi type token.
 	pkg, module, member, diagnostics := r.DecomposeToken()
+	if pkg == "pulumi" && module == "pulumi" {
+		module = ""
+	}
```

Comment:
Should we refactor this out to a top level method alongside `MapProvidersAsResources`? We need to do this for each language, may as well export it as a method. If we add any future special cases & built-in methods, that'll be the place to do it.

Reply:
For collecting imports, I think it may still be helpful to still have `mod` information to distinguish between needing to import `pulumi:Providers:<name>` versus ignoring `pulumi:pulumi:*`, like for [here](https://github.com/pulumi/pulumi/pull/10419/files#diff-d6817c1c54a334e663ae7a8ef2bcc697641fcddf412b0d2b97c884a9e297a5f3R364)

Reply:
I think you've found some dead code since I introduced `MapProvidersAsResources`. We'll have already mapped `pulumi:providers:$name` to `$name::Provider`, so I don't think we'll ever end up in this branch:

```
if pkg == "pulumi" && mod == "providers" {
```

Let me put my thinking another way: I'd like any sort of "special case" logic about what the `pulumi` package does that _isn't_ like other providers to live in `codegen/pcl/utilities.go` because eventually we will want language plugins & codegen to be split out into separate repos, and we won't want to have to update N repos with new special cases. (e.g.: Pulumi-Java)

If we move the mapping here into a `func FixupPulumiPackageTokens` that's shared, then we can update that in one place and as our language plugins update, we don't need to find and update the same thing in each of them.

Reply:
Moved the logic to reorg pulumi package tokens to `utilities.go`; however, we still currently special case ignoring importing the `pulumi` package since `genpreamble` does that already, and it attempts to do it incorrectly i.e. `pulumi_pulumi` (the original issue from the ticket)

Reply:
That makes sense.